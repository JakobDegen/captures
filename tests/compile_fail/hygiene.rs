use captures::*;

// Check that `capture_only` does what we expected
fn basic() {
    let a = 1;
    let b = 2;
    let f = capture_only!(clone a, move || {
        let mut total = 0;
        total += a;
        total += b;
        total
    });
}

fn main() {
    basic();
}
