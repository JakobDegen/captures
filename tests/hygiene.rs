use captures::*;

// Don't wipe the surrounding context
fn context() {
    const FOO: i32 = 0;
    struct Bar;
    fn func() {}

    let mut a = Bar;
    let mut f = capture_only!(all a, || {
        if FOO > 1 {
            a = Bar;
            func();
        }
    });
    f();
}

// we can shadow things
fn shadow() {
    let mut a = 5;
    let mut f = capture_only!(all a, || {
        a += 1;
        a = { // 9
            let a = a;
            a + 3
        };
        a = if let a @ 10..=15 = a {
            a + 3
        } else {
            a
        };
        a = match a {
            a @ 9 => 100,
            a => 50
        };
    });
    f();
    assert_eq!(a, 100);
}

fn main() {
    context();
    shadow();
}
