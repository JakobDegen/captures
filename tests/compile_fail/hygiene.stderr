error[E0425]: cannot find value `b` in this scope
  --> tests/compile_fail/hygiene.rs:10:18
   |
10 |         total += b;
   |                  ^ not found in this scope
   |
   = note: this error originates in the macro `capture_only` (in Nightly builds, run with -Z macro-backtrace for more info)
